### PR TITLE
Fast DC security analysis using Woodbury engine

### PR DESCRIPTION
**Pull request by p-arvy**
**Please check if the PR fulfills these requirements**

- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem?**

No.

**What kind of change does this PR introduce?**

A fast DC SA based on Woodbury equality is added. **Please note that for now, no action is supported in this SA, and that contingencies on switches are not yet supported.**

To test it, and to check that it is isofunctional with the current DC SA implementation (actions/contingencies on switches aside), the same TUs from the `OpenSecurityAnalysisTest` class are used. To do so, TUs are parameterized to be run with both DC AS implementations, using the boolean parameter `dcFastMode` from the `OpenSecurityAnalysisParameter` class.

When it is necessary, TUs are split so there is no redundancy in what is being tested (e.g. when an AC AS and a DC AS are run in the same TU).

**What is the current behavior?**

No such SA exists.

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change or deprecate an API?**
- [ ] Yes
- [ ] No

**If yes, please check if the following requirements are fulfilled**

- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration steps are described in the following section

**What changes might users need to make in their application due to this PR? (migration steps)**

**Other information**:

### CODE REVIEW

File: src/main/java/com/powsybl/openloadflow/sa/WoodburyDcSecurityAnalysis.java [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+                .flatMap(contingency -> contingency.getBranchIdsToOpen().keySet().stream())
+                .map(branchId -> lfNetwork.getBranchById(branchId).getBranchType())
+                .filter(branchType -> branchType == LfBranch.BranchType.SWITCH)
+                .findAny()
+                .ifPresent(e -> {
+                    throw new IllegalArgumentException("Contingencies on switch not yet supported in fast DC Security Analysis");
+                });
+
+        // map contingencies on buses to contingencies on linked branches terminals
+        propagatedContingencies.stream()
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/sa/WoodburyDcSecurityAnalysis.java:163]: I think this is not what you have to do. You have to used `cleanContingencies` from DcSensitivityAnalysis. Right? If yes, I suggest to put it as utility method in PropagatedContingency.

File: src/main/java/com/powsybl/openloadflow/sa/WoodburyDcSecurityAnalysis.java [hunk 2]
```diff
@@ -89,8 +90,7 @@ private DenseMatrix calculatePostContingencyStatesForAContingency(DcLoadFlowCont
 
             // if a phase tap changer is lost or if the connectivity have changed, we must recompute load flows
             if (!disabledBuses.isEmpty() || !lostPhaseControllers.isEmpty()) {
-                double[] newFlowStatesArray = DcSensitivityAnalysis.runDcLoadFlow(loadFlowContext, disabledNetwork, reportNode);
-                newFlowStates = new DenseMatrix(newFlowStatesArray.length, 1, newFlowStatesArray);
+                newFlowStates = DcSensitivityAnalysis.runDcLoadFlow(loadFlowContext, disabledNetwork, reportNode);
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/sa/WoodburyDcSecurityAnalysis.java:93]: This dependency to DcSensitivityAnalysis is not allowed.

File: src/main/java/com/powsybl/openloadflow/sensi/DcSensitivityAnalysis.java [hunk 3]
```diff
@@ -388,7 +388,8 @@ private void processContingenciesBreakingConnectivity(ConnectivityBreakAnalysis.
                 connectivityAnalysisResult.getElementsToReconnect(), resultWriter, reportNode, partialDisabledBranches, rhsChanged);
     }
 
-    protected void cleanContingencies(LfNetwork lfNetwork, List<PropagatedContingency> contingencies) {
+    // TODO : move this method in PropagatedContingencies in another PR
+    public static void cleanContingencies(LfNetwork lfNetwork, List<PropagatedContingency> contingencies) {
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/sensi/DcSensitivityAnalysis.java:392]: Not allowed, to be fixed this that PR or in a PR before.

File: src/test/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysisTest.java [hunk 4]
```diff
@@ -1921,7 +2006,31 @@ void testDcBusBarSectionContingency() {
     }
 
     @Test
-    void testBusBarSectionContingencyWithSlackBusRelocation() {
+    void testDcBusBarSectionContingencyFastMode() {
```
Comment (Hadrien-Godard) [inline src/test/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysisTest.java:2009]: This test should be reworked if contingencies on switches are handled in this PR.

File: src/main/java/com/powsybl/openloadflow/sensi/DcSensitivityAnalysis.java (hunk 3 above)
Comment (annetill) [reply src/main/java/com/powsybl/openloadflow/sensi/DcSensitivityAnalysis.java:392]: @p-arvy what about this issue?

Comment (p-arvy) [reply src/main/java/com/powsybl/openloadflow/sensi/DcSensitivityAnalysis.java:392]: Thank for getting back to this @annetill, I opened a PR dedicated to this issue: https://github.com/powsybl/powsybl-open-loadflow/pull/1081

File: src/main/java/com/powsybl/openloadflow/sa/WoodburyDcSecurityAnalysis.java (hunk 2 above)
Comment (p-arvy) [reply src/main/java/com/powsybl/openloadflow/sa/WoodburyDcSecurityAnalysis.java:93]: Same, here is a PR dedicated to this issue: https://github.com/powsybl/powsybl-open-loadflow/pull/1082

File: src/main/java/com/powsybl/openloadflow/sa/WoodburyDcSecurityAnalysis.java [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+                                    connectivityBreakAnalysisResults.contingencyElementByBranch(), Collections.emptySet(), Collections.emptySet(), reportNode, Collections.emptySet());
+                            PostContingencyResult postContingencyResult = computePostContingencyResult(context, securityAnalysisParameters, preContingencyLimitViolationManager,
+                                    preContingencyNetworkResult, contingency, postContingencyStates, limitReductions, createResultExtension);
+                            postContingencyResults.add(postContingencyResult);
+                            networkState.restore();
+                        }
+                    });
+
+            LOGGER.info("Processing post contingency results for contingencies breaking connectivity");
+            connectivityBreakAnalysisResults.connectivityAnalysisResults()
```
Comment (vidaldid-rte) [inline src/main/java/com/powsybl/openloadflow/sa/WoodburyDcSecurityAnalysis.java:248]: does connectivityAnalysisResult return a connectivity result for the contingncis that break connectivity ?
If yes, the naming is not so great.

File: src/test/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysisTest.java (hunk 4 above)
Comment (vidaldid-rte) [reply src/test/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysisTest.java:2009]: why testDcBusBarSectionContingency isn't a parametrized test with fastMode true/false ?

File: src/main/java/com/powsybl/openloadflow/sa/WoodburyDcSecurityAnalysis.java (hunk 1 above)
Comment (Hadrien-Godard) [reply src/main/java/com/powsybl/openloadflow/sa/WoodburyDcSecurityAnalysis.java]: Work in progress in PR #1081.

File: src/test/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysisTest.java (hunk 4 above)
Comment (Hadrien-Godard) [reply src/test/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysisTest.java:2009]: Currently contingencies on switches are not handled, but we are working on it. When done, this test will be parametrized.

Comment (p-arvy) [reply src/test/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysisTest.java]: Contingencies are now handled, and the TU is parameterized. Thank you for the comments. I resolve this conversation.

File: src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysisProvider.java [hunk 6]
```diff
@@ -82,7 +82,11 @@ public CompletableFuture<SecurityAnalysisReport> run(Network network,
 
         AbstractSecurityAnalysis<?, ?, ?, ?, ?> securityAnalysis;
         if (loadFlowParameters.isDc()) {
-            securityAnalysis = new DcSecurityAnalysis(network, matrixFactory, selectedConnectivityFactory, runParameters.getMonitors(), runParameters.getReportNode());
+            if (OpenSecurityAnalysisParameters.getOrDefault(runParameters.getSecurityAnalysisParameters()).isDcFastMode()) {
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysisProvider.java:85]: @obrix @jeandemanged can you check this line about security analysis parameters please ! I am not familiar with them.

File: README.md [hunk 7]
```diff
@@ -69,6 +69,7 @@ Heterogeneous voltage controls management has become a key feature. All well-mod
  - Complex cases where the contingency leads to another synchronous component where a new resolution has to be performed are not supported at that stage. The loss of slack bus during a contingency is not supported yet, but the work is in progress.
  - The active and reactive power flows on branches, as well as angle and voltage at buses, can be monitored and collected for later analysis after the base case and after each contingency.
  - Remedial actions such as: switch action, terminal(s) connection action, re-dispatching action
+ - Fast DC mode available, based on Woodbury's formula for calculating post-contingency states. Note that this mode has limitations for the moment, such as remedial actions not yet taken into account.
```
Comment (annetill) [inline README.md:72]: Can you be more precise about management of HVDC in AC emulation?

File: docs/security/parameters.md [hunk 8]
```diff
@@ -50,6 +50,17 @@ The thread pool used for getting threads is the one provided by the `Computation
...
 The default value is 1.
 
+**dcFastMode**  
+The `dcFastMode` property allows to use fast DC security analysis, based on Woodbury's formula for calculating post-contingency states, 
+when DC mode is activated.
+
+Please note that fast mode has a few limitations:
+- Contingencies applied on branches opened on one side are ignored.
+- No remedial action is currently supported.
+- Slack relocation following the application of a contingency is not supported.
```
Comment (annetill) [inline docs/security/parameters.md:60]: And HVDCs?

File: src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysisProvider.java (hunk 6 above)
Comment (annetill) [reply src/main/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysisProvider.java:85]: To be fixed in another PR.

File: README.md (hunk 7 above)
Comment (Hadrien-Godard) [reply README.md]: Done

File: docs/security/parameters.md (hunk 8 above)
Comment (Hadrien-Godard) [reply docs/security/parameters.md:61]: Done

**Review by Hadrien-Godard — APPROVED** [review]

**Review by annetill — APPROVED** [review]
Looks good, I'am happy, it was not easy at all :-)

**Pull request merged**